psi4/src/psi4/libmints/basisset.cc

```diff
@@ -192,23 +211,51 @@ int BasisSet::n_frozen_core(const std::string &depth, SharedMolecule mol) {
             // Exclude ghosted atoms from core-freezing
             if (Z > 0) {
                 // Add ECPs to Z, the number of electrons less ECP-treated electrons
-                double ECP = n_ecp_core(mymol->label(A));
-                if (Z + ECP > 2) nfzc += 1;
-                if (Z + ECP > 10) nfzc += 4;
-                if (Z + ECP > 18) nfzc += 4;
-                if (Z + ECP > 36) nfzc += 9;
-                if (Z + ECP > 54) nfzc += 9;
-                if (Z + ECP > 86) nfzc += 16;
-                if (Z + ECP > 108) {
-                    throw PSIEXCEPTION("Invalid atomic number");
+                int ECP = n_ecp_core(mymol->label(A));
+                int current_shell = atom_to_period(Z + ECP);
+                int delta = period_to_full_shell(current_shell - 1);
+                // Keep track of the largest frozen shell, in case its a cationic species
+                if (largest_shell < current_shell) {
+                    largest_shell = current_shell;
                 }
-                // If this center has an ECP, some pairs are already frozen
-                if (ECP > 0) nfzc -= ECP / 2;
+                // If this center has an ECP, some electrons are already frozen
+                if (ECP > 0) delta -= ECP;
+                // Keep track of current valence electrons
+                mol_valence = mol_valence + Z - delta;
+                num_frozen_el += delta;
             }
         }
-        return nfzc;
+        // If we are about to end up with no valence electrons,
+        // unfreeze electrons from the largest shell in the molecule
+        if (mol_valence <= 0) num_frozen_el -= period_to_full_shell(largest_shell - 1) - period_to_full_shell(largest_shell - 2);
+        return num_frozen_el/2;
     } else {
-        throw std::invalid_argument("Frozen core spec is not supported, options are {true, false}.");
+        // Options are filtered in read_options.cc; allowed strings are:
+        // TRUE, FALSE, -1, -2, -3
+        int req_shell = -std::stoi(local, nullptr, 10);
+        int num_frozen_el = 0;
+        int mol_valence = -1 * mymol->molecular_charge();
+        // Freeze the number of core electrons strictly corresponding to the
+        // requested previous n-th shell.
+        for (int A = 0; A < mymol->natom(); A++) {
+            double Z = mymol->Z(A);
```

Comment:
Isn't this an int?

Reply:
Doesn't look like that to me:

https://github.com/psi4/psi4/blob/9e6329abb493ed79dbb6913ac4c1df957e9839ae/psi4/src/psi4/libmints/molecule.h#L255-L256

Reply:
Huh... why is that

Reply:
Because in the beginning there was the thought that `Z` would be effective nuclear charge and float in case any MM atoms were around. At least that's my understanding.